### ISSUE TITLE
Stop vendoring runc / libcontainer

### ISSUE THREAD

**Issue by saschagrunert** [kubernetes/kubernetes#104325]
#### What happened:

Kubernetes vendors runc, for example to cleanup stale cgroups.

https://github.com/kubernetes/kubernetes/blob/e7efe219f8a5fcdff9647d45639696048c1f6147/go.mod#L367

#### What you expected to happen:

We should not rely on lower level libraries like libcontainer, because they do not guarantee anything in terms of stability. The only valid interface of lower level OCI runtimes is their CLI. Updating runc in a cluster already has impact in the whole stack and from my point of view we should maintain parallel implementations rather than having to many moving parts.

#### Anything else we need to know?:

Ref: https://github.com/kubernetes/kubernetes/issues/104280

**Comment by dims** [comment]
💯 @saschagrunert

**Comment by odinuge** [comment]
Thanks for creating this issue @saschagrunert! We (I have at least suggested this quite a few times :sweat_smile:) have been discussing this quite a lot lately, so it is nice to see that this others agree about the issue!

I think we can reduce the overhead of kubelet quite a fair amout from implementing this ourselves. We don't use much of the logic in runc, but its impact is huge. We have seen a fair amount of bugs in kubelet caused by seemingly unrelated changes in runc, essentially breaking kubelet for all users. Also, since runc has a different objective than k8s, it becomes hard to manage them "together" as we do now. Often the discussion and changes in runc are not favorable for k8s (both in terms of overhead and complexity), while they are necessary for runc itself. All this making it even harder to manage; both for us/k8s and the runc maintainers.

Another big issue is that since we only use a small subset of runc, reviewing big changes in its codebase is quite difficult; especially since we are lacking CI coverage for the systemd cgroup driver.

Implementing this ourselves also gives us a lot more flexibility when it comes to unit testing things for _our use_, as well as making reviewing a lot easier. I don't vote for a "complete rewrite PR", but something more gradual would be nice! We might keep using some code `runc`, but rewrite the small cgroup enforcement we use ourselves. I will be more than happy helping to eg. write a KEP about the way forward, together with its pros and cons. Looking forward to hearing other thoughts about this!

**Comment by kolyshkin** [comment]
This is more or less inline with what I had in mind. Unfortunately, runc/libcontainer is the only library which has the functionality of creating transient systemd scope units -- and this is the functionality that is not used by runc itself, as it was added specifically for k8s. Surely, this functionality was totally untested (although we added a few test cases here and there recently).

Unless there's some other library that can be used for what runc/libcontainer/cgroups/systemd is used by kubelet's cgroup manager, I think the way to proceed is to separate the common systemd handling logic (dbus connection wrapper, various helper functions etc) out of runc/libcontainer/cgroups/systemd into a separate package (say `opencontainers/systemd`, doesn't really matter) and reuse it from both runc and kubernetes.

I think it's important to avoid doing mere copy/paste and rewrite -- this usually works wonders in a short-term perspective, but creates a lot of technical debt in the long run (I have seen oh so many examples of that).

I can start working on this once I wrap up a few things.
